image/qcow2/qcow2.go

```diff
@@ -959,6 +959,122 @@ func readZero(p []byte, off int64, sz uint64) (int, error) {
+	// return an unallocated cluster.
+	if img.backingImage == nil || off >= img.backingImage.Size() {
+		return unallocated, nil
+	}
+
+	// Get the cluster from the backing file.
+	length := int64(img.clusterSize)
+	if off+length > img.backingImage.Size() {
+		length = img.backingImage.Size() - off
+	}
+	parent, err := img.backingImage.Next(off, length)
+	if err != nil {
+		return parent, err
+	}
+	// The backing image may be a raw image not aligned to cluster size.
+	parent.Length = int64(img.clusterSize)
+	return parent, nil
+}
+
+// clusterStatus returns an extent describing a single cluster. off must be aligned to
+// cluster size.
+func (img *Qcow2) clusterStatus(off int64) (image.Extent, error) {
+	status := image.Extent{Start: off, Length: int64(img.clusterSize)}
+	var cm clusterMeta
+	if err := img.getClusterMeta(off, &cm); err != nil {
+		return status, err
+	}
+	if !cm.Allocated {
+		return img.clusterStatusUnallocated(off)
+	}
+	// Cluster present in this image.
+	status.Allocated = true
+	desc := cm.L2Entry.clusterDescriptor()
+	if cm.Compressed {
+		status.Compressed = true
+	} else {
+		standardDesc := standardClusterDescriptor(desc)
+		if img.extendedL2() {
+			// TODO: report entire cluster as data until we understand better the
+			// semantics of extended l2 clusters. This is correct but inefficient.
+			status.Zero = false
+		} else {
+			status.Zero = standardDesc.allZero()
+		}
+	}
+	return status, nil
+}
+
+// Return true if extents have the same status.
+func sameStatus(a, b image.Extent) bool {
+	return a.Allocated == b.Allocated && a.Zero == b.Zero && a.Compressed == b.Compressed
+}
+
+// Next returns the next extent in this image. An extent describes one or more
+// clusters having the same status.
+func (img *Qcow2) Next(start, length int64) (image.Extent, error) {
+	// Default to zero length non-existent cluster, read as zero.
+	var current image.Extent
+
+	if img.errUnreadable != nil {
+		return current, img.errUnreadable
+	}
+	if img.clusterSize == 0 {
+		return current, errors.New("cluster size cannot be 0")
+	}
+
+	// Fail out of bounds requests. User can use Size() to ensure requests are
+	// within the image.
+	if start > int64(img.Header.Size) {
+		return current, errors.New("start out of bounds")
+	}
+	if start+length > int64(img.Header.Size) {
+		return current, errors.New("length out of bounds")
+	}
+
+	offset := start
+	remaining := length
+
+	for remaining > 0 {
+		clusterNo := offset / int64(img.clusterSize)
+		clusterStart := clusterNo * int64(img.clusterSize)
+		clusterStatus, err := img.clusterStatus(clusterStart)
+		if err != nil {
+			return current, err
+		}
+
+		// If start+length is not aligned to cluster size, fixup the length.
+		if clusterStatus.Length > remaining {
+			clusterStatus.Length = remaining
+		}
+
+		if current.Length == 0 {
+			// First cluster: copy status to current.
+			current = clusterStatus
+			current.Start = start
```

Comment:
If start was not aligned to cluster size need to fixup current.Length to remove start-clusterStart.